Source/Core/Core/HW/WiiSave.cpp

```diff
@@ -41,525 +48,509 @@ constexpr Md5 s_md5_blanker{{0x0E, 0x65, 0x37, 0x81, 0x99, 0xBE, 0x45, 0x17, 0xA
+    ScanForFiles();
   }
-  fpData_bin.Close();
 
-  if (m_bk_hdr.size != BK_LISTED_SZ || m_bk_hdr.magic != BK_HDR_MAGIC)
-  {
-    ERROR_LOG(CONSOLE, "Invalid Size(%x) or Magic word (%x)", u32(m_bk_hdr.size),
-              u32(m_bk_hdr.magic));
-    m_valid = false;
-    return;
-  }
+  bool SaveExists() override { return File::Exists(m_wii_title_path + "/banner.bin"); }
 
-  if (m_bk_hdr.size_of_files + FULL_CERT_SZ != m_bk_hdr.total_size)
+  std::optional<Header> ReadHeader() override
   {
-    WARN_LOG(CONSOLE, "Size(%x) + cert(%x) does not equal totalsize(%x)",
-             u32(m_bk_hdr.size_of_files), FULL_CERT_SZ, u32(m_bk_hdr.total_size));
+    Header header{};
+    std::string banner_file_path = m_wii_title_path + "/banner.bin";
+    u32 banner_size = static_cast<u32>(File::GetSize(banner_file_path));
+    header.hdr.banner_size = banner_size;
+    header.hdr.tid = m_tid;
+    header.hdr.md5 = s_md5_blanker;
+    header.hdr.permissions = 0x3C;
+
+    File::IOFile banner_file(banner_file_path, "rb");
+    if (!banner_file.ReadBytes(header.banner, banner_size))
+      return {};
+    // remove nocopy flag
+    header.banner[7] &= ~1;
+
+    Md5 md5_calc;
+    mbedtls_md5(reinterpret_cast<const u8*>(&header), sizeof(Header), md5_calc.data());
+    header.hdr.md5 = std::move(md5_calc);
+    return header;
   }
-  if (m_title_id != m_bk_hdr.save_game_title)
+
+  std::optional<BkHeader> ReadBkHeader() override
   {
-    WARN_LOG(CONSOLE,
-             "Encrypted title (%" PRIx64 ") does not match unencrypted title (%" PRIx64 ")",
-             m_title_id, u64(m_bk_hdr.save_game_title));
+    BkHeader bk_hdr{};
+    bk_hdr.size = BK_LISTED_SZ;
+    bk_hdr.magic = BK_HDR_MAGIC;
+    bk_hdr.ngid = s_ng_id;
+    bk_hdr.number_of_files = static_cast<u32>(m_files_list.size());
+    bk_hdr.size_of_files = m_files_size;
+    bk_hdr.total_size = m_files_size + FULL_CERT_SZ;
+    bk_hdr.tid = m_tid;
+    return bk_hdr;
   }
-}
 
-void WiiSave::WriteBKHDR()
-{
-  if (!m_valid)
-    return;
-  u32 number_of_files = 0, size_of_files = 0;
-  ScanForFiles(m_wii_title_path, m_files_list, &number_of_files, &size_of_files);
-  memset(&m_bk_hdr, 0, BK_SZ);
-  m_bk_hdr.size = BK_LISTED_SZ;
-  m_bk_hdr.magic = BK_HDR_MAGIC;
-  m_bk_hdr.ngid = s_ng_id;
-  m_bk_hdr.number_of_files = number_of_files;
-  m_bk_hdr.size_of_files = size_of_files;
-  m_bk_hdr.total_size = size_of_files + FULL_CERT_SZ;
-  m_bk_hdr.save_game_title = m_title_id;
-
-  File::IOFile data_file(m_encrypted_save_path, "ab");
-  if (!data_file.WriteBytes(&m_bk_hdr, BK_SZ))
+  std::optional<std::vector<SaveFile>> ReadFiles() override
   {
-    ERROR_LOG(CONSOLE, "Failed to write bkhdr");
-    m_valid = false;
+    std::vector<SaveFile> ret(m_files_list.size());
+    std::transform(m_files_list.begin(), m_files_list.end(), ret.begin(), [this](const auto& path) {
+      const File::FileInfo file_info{path};
+      SaveFile save_file;
+      save_file.mode = 0x3c;
```

Comment:
Do we have a (combination of) constant(s) for this one anywhere? Or at least comment on why it is that value (`0'074` to put it in oct, as most chmod commands use) instead of any other - assuming you know; since the old code also had it and you might've just copied it over.

Reply:
Sure, I just pointed out magic numbers whose meaning will probably be lost to time in a week or so :P
If it's scheduled to go away, I'm fine with that.